### PR TITLE
test: Add cases about proposed/committed rewards

### PR DESCRIPTION
**Pull request by keroro520**
* test: Add cases about proposed/committed rewards
* test: Add forked_transaction
* feat(test): Provide a util function to generating live cells
* feat(test): Add TXOSet to manage cells
* chore: Export all test cases in specs

### CODE REVIEW

File: test/src/specs/mining/fee.rs [hunk 1]
```diff
@@ -1 +1,355 @@
...
+        let finalization_delay_length = node.consensus().finalization_delay_length();
+
+        let txs = generate_utxo_set(node, 1).bang_random_fee(vec![node.always_success_cell_dep()]);
+        node.submit_transaction(&txs[0]);
+
+        let number_to_submit = node.get_tip_block_number();
+        let number_to_propose = number_to_submit + 1;
+        let number_to_commit = number_to_propose + closest;
+        node.generate_blocks(2 * finalization_delay_length as usize);
+        assert_block(node, number_to_propose, &txs, &[]);
```
Comment (u2) [inline test/src/specs/mining/fee.rs]: Suggest to use `assert_proposals`, `assert_committed` directly, `assert_block` is obscure. The same as `assert_chain`

File: test/src/specs/mining/fee.rs [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+}
+
+pub struct ProposeButNotCommit;
+
+impl Spec for ProposeButNotCommit {
+    crate::name!("propose_but_not_commit");
+
+    crate::setup!(num_nodes: 2, connect_all: false);
+
+    // Case: Propose a transaction but never commit it
```
Comment (u2) [inline test/src/specs/mining/fee.rs]: Don't get this test case. How does the node not commit it?

File: test/src/specs/sync/chain_forks.rs [hunk 3]
```diff
@@ -552,6 +553,63 @@ impl Spec for ForksContainSameTransactions {
...
+        let node0 = &net.nodes[0];
+        let node1 = &net.nodes[1];
+        let finalization_delay_length = node0.consensus().finalization_delay_length();
+
+        net.exit_ibd_mode();
+        let fixed_point = node0.get_tip_block_number();
+        let tx = node1.new_transaction_spend_tip_cellbase();
+
+        // node0 doesn't have `tx`      => TxStatus: None
+        // node1 have `tx` on main-fork => TxStatus: Some(Committed)
```
Comment (u2) [inline test/src/specs/sync/chain_forks.rs]: the main-fork is the main-chain?

File: test/src/specs/sync/chain_forks.rs [hunk 4]
```diff
@@ -552,6 +553,63 @@ impl Spec for ForksContainSameTransactions {
...
+
+            let tx_status = node0.rpc_client().get_transaction(tx.hash());
+            assert!(tx_status.is_none(), "node0 maintains tx in unverified fork");
+            let is_pending = |tx_status: &TransactionWithStatus| {
+                let pending_status = TxStatus::pending();
+                tx_status.tx_status.status == pending_status.status
+            };
+            let tx_status = node1.rpc_client().get_transaction(tx.hash()).unwrap();
+            assert!(
+                is_pending(&tx_status),
```
Comment (u2) [inline test/src/specs/sync/chain_forks.rs]: I not sure why the tx_status changes from `Committed` to `Pending`.

File: test/src/specs/sync/chain_forks.rs (hunk 3 above)
Comment (keroro520) [reply test/src/specs/sync/chain_forks.rs]: Thanks for taking the time to seriously review the test-cases. I have added some code comments to explain the process of test-cases. Please take a look again.

And `assert_block` and `assert_chain` has been removed based on your suggestions.

**Comment by u2** [comment]
Please fix the conflict.

**Pull request merged**